### PR TITLE
gtk: Automatically generate ParamSpecExpression

### PR DESCRIPTION
**Pull request by A6GibKm**

### CODE REVIEW

File: gtk4/src/param_spec_expression.rs [hunk 1]
```diff
@@ -1,26 +1,11 @@
 // Take a look at the license at the top of the repository in the LICENSE file.
 
+use crate::ParamSpecExpression;
+
 use glib::gobject_ffi;
 use glib::translate::*;
 use glib::ParamSpec;
 
-glib::wrapper! {
-    #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
```
Comment (bilelmoussaoui) [inline gtk4/src/param_spec_expression.rs:8]: you forgot to remove it from lib.rs

Comment (bilelmoussaoui) [reply gtk4/src/param_spec_expression.rs:8]: Also copy the value trait implementations from https://github.com/gtk-rs/gtk-rs-core/blob/master/glib/src/param_spec.rs#L37-L84  those were wrongly implemented in the past

File: gtk4/Gir.toml [hunk 2]
```diff
@@ -1606,6 +1605,12 @@ status = "generate"
     pattern = "*-set" # those properties are read only
     generate = ["get", "notify"]
 
+[[object]]
+name = "Gtk.ParamSpecExpression"
+status = "generate"
+manual = true
```
Comment (bilelmoussaoui) [inline gtk4/Gir.toml:1611]:
```suggestion
```
not sure where you got this one from but it is not needed

File: gtk4/src/param_spec_expression.rs (hunk 1 above)
Comment (bilelmoussaoui) [inline gtk4/src/param_spec_expression.rs:8]: As those derives are not auto-generated for shared types, i would at least implement Debug manually

Comment (A6GibKm) [reply gtk4/src/param_spec_expression.rs:8]: How do I add those traits?

Comment (bilelmoussaoui) [reply gtk4/src/param_spec_expression.rs:8]: just add the debug one, copy paste the automatically generate Display and replace Display with Debug :)

File: gtk4/src/param_spec_expression.rs [hunk 3]
```diff
@@ -62,6 +53,55 @@ impl ParamSpecExpression {
     }
 }
 
+#[doc(hidden)]
+impl glib::value::ValueType for ParamSpecExpression {
+    type Type = ParamSpec;
```
Comment (bilelmoussaoui) [inline gtk4/src/param_spec_expression.rs:58]:
```suggestion
    type Type = ParamSpecExpression;
```

File: gtk4/src/param_spec_expression.rs [hunk 4]
```diff
@@ -62,6 +53,55 @@ impl ParamSpecExpression {
...
+        assert!(!ptr.is_null());
+        from_glib_full(ptr as *mut gobject_ffi::GParamSpec)
+    }
+}
+
+#[doc(hidden)]
+impl glib::value::ToValue for ParamSpecExpression {
+    fn to_value(&self) -> Value {
+        unsafe {
+            let mut value = Value::from_type(ParamSpecExpression::static_type());
```
Comment (bilelmoussaoui) [inline gtk4/src/param_spec_expression.rs:79]:
```suggestion
            let mut value = Value::from_type(Self::static_type());
```

**Review by bilelmoussaoui — CHANGES_REQUESTED** [review]

**Comment by bilelmoussaoui** [comment]
Thank you!

**Pull request merged**